### DISCUSSION TITLE
Autocomplete when instantiating an ORM class

### DISCUSSION THREAD

**Discussion by CozyDoomer** [sqlalchemy/sqlalchemy#10359] [Usage Questions] [answered]
If I create a table using the ORM like this:
```python
class Base(DeclarativeBase):
    pass

class TestTable1(Base):
    __tablename__ = "test_table1"
    test_id: Mapped[int] = mapped_column(Integer, primary_key=True)
    test_name: Mapped[str] = mapped_column(String)
    test_version: Mapped[str] = mapped_column(String)
```
I don't see any information about the attributes:
![image](https://github.com/sqlalchemy/sqlalchemy/assets/41551012/a4fbe252-9178-472b-853f-a2b44e50b0a9)

This is also true for autocomplete - I don't see any of the attributes being suggested when instantiating the ORM class.  
- - -

Only if I explicitly add a constructor like this:
```python
class TestTable2(Base):
    def __init__(self, test_name: str, test_version: str):
        super().__init__(test_name, test_version)

    __tablename__ = "test_table2"
    test_id: Mapped[int] = mapped_column(Integer, primary_key=True)
    test_name: Mapped[str] = mapped_column(String)
    test_version: Mapped[str] = mapped_column(String)
```

I get information about the attributes:
![image](https://github.com/sqlalchemy/sqlalchemy/assets/41551012/2b0c0e00-fd22-4367-b338-8386e55f1b33)

- - -

Is there a way to get information about the attributes before/during instantiation without explicitly defining a constructor?

**Comment by zzzeek** [comment, marked as answer]
the [default constructor](https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#default-constructor) accepts only ``**kw`` and is not generated from specific mapped attributes.

to create a constructor that is very specific to mapped attributes, use the [dataclasses](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping) feature.

**Reply by CozyDoomer** [reply]
Thank you for the quick response!

Which dataclasses feature would generate a default constructor that shows the attribute names as positional and keyword arguments?
I'm still not seeing the attributes using the User example from the docs.

```python
class Base(MappedAsDataclass, DeclarativeBase):
    """subclasses will be converted to dataclasses"""

class User(Base):
    __tablename__ = "user_account"

    id: Mapped[int] = mapped_column(init=False, primary_key=True)
    name: Mapped[str]
```

(switched to jupyter notebook to make sure it's not a lsp issue)   

![image](https://github.com/sqlalchemy/sqlalchemy/assets/41551012/f4fb17db-a04c-4da9-95f8-f010bcb220b5)

The other feature I aimed for `__repr__` seems to work fine though:
![image](https://github.com/sqlalchemy/sqlalchemy/assets/41551012/61f2a965-d08d-4672-936a-d9eee90e7e9b)

**Reply by CaselIT** [reply]
that seems to be a jupyter issue. Running
```py
from sqlalchemy.orm import MappedAsDataclass, DeclarativeBase, mapped_column, Mapped

class Base(MappedAsDataclass, DeclarativeBase):
    """subclasses will be converted to dataclasses"""

class User(Base):
    __tablename__ = "user_account"

    id: Mapped[int] = mapped_column(init=False, primary_key=True)
    name: Mapped[str]

help(User)
```
shows the correct init
```
Help on class User in module __main__:

class User(Base)
 |  User(name)
 |
 |  User(name: str)
 |
 |  Method resolution order:
 |      User
 |      Base
 |      sqlalchemy.orm.decl_api.MappedAsDataclass
 |      sqlalchemy.orm.decl_api.DeclarativeBase
 |      sqlalchemy.inspection.Inspectable
 |      typing.Generic
 |      builtins.object
 |
 |  Methods defined here:
 |
 |  __eq__(self, other)
 |      Return self==value.
 |
 |  __init__(self, name)
 |
 |  __repr__(self)
 |      Return repr(self).
 |
 |  ----------------------------------------------------------------------
 |  Data descriptors defined here:
 |
 |  id
 |
 |  name
 |
 |  ----------------------------------------------------------------------
 |  Data and other attributes defined here:
 |
 |  __annotations__ = {'id': sqlalchemy.orm.base.Mapped[int], 'name': sqla...
 |
 |  __dataclass_fields__ = {'id': Field(name='id',type=<class 'int'>,defau...
 |
 |  __dataclass_params__ = _DataclassParams(init=True,repr=True,eq=True,or...
 |
 |  __hash__ = None
 |
 |  __mapper__ = <Mapper at 0x2cb24c5ff10; User>
 |
 |  __match_args__ = ('name',)
 |
 |  __parameters__ = ()
 |
 |  __table__ = Table('user_account', MetaData(), Column('id', I...ble=<us...
 |
 |  __tablename__ = 'user_account'
 |
 |  ----------------------------------------------------------------------
 |  Data and other attributes inherited from Base:
 |
 |  metadata = MetaData()
 |
 |  registry = <sqlalchemy.orm.decl_api.registry object>
 |
 |  ----------------------------------------------------------------------
 |  Class methods inherited from sqlalchemy.orm.decl_api.MappedAsDataclass:
 |
 |  __init_subclass__(init: 'Union[_NoArg, bool]' = _NoArg.NO_ARG, repr: 'Union[_NoArg, bool]' = _NoArg.NO_ARG, eq: 'Union[_NoArg, bool]' = _NoArg.NO_ARG, order: 'Union[_NoArg, bool]' = _NoArg.NO_ARG, unsafe_hash: 'Union[_NoArg, bool]' = _NoArg.NO_ARG, match_args: 'Union[_NoArg, bool]' = _NoArg.NO_ARG, kw_only: 'Union[_NoArg, bool]' = _NoArg.NO_ARG, dataclass_callable: 'Union[_NoArg, Callable[..., Type[Any]]]' = _NoArg.NO_ARG) -> 'None' from sqlalchemy.orm.decl_api.DCTransformDeclarative
 |      This method is called when a class is subclassed.
 |
 |      The default implementation does nothing. It may be
 |      overridden to extend subclasses.
 |
 |  ----------------------------------------------------------------------
 |  Data descriptors inherited from sqlalchemy.orm.decl_api.MappedAsDataclass:
 |
 |  __dict__
 |      dictionary for instance variables (if defined)
 |
 |  __weakref__
 |      list of weak references to the object (if defined)
 |
 |  ----------------------------------------------------------------------
 |  Data and other attributes inherited from sqlalchemy.orm.decl_api.DeclarativeBase:
 |
 |  __orig_bases__ = (sqlalchemy.inspection.Inspectable[sqlalchemy.orm.sta...
 |
 |  ----------------------------------------------------------------------
 |  Class methods inherited from typing.Generic:
 |
 |  __class_getitem__(params) from sqlalchemy.orm.decl_api.DCTransformDeclarative
```

**Reply by CozyDoomer** [reply]
Jupyter issue seemed very plausible at first but I couldn't get it to work in other editors - also when using different LSPs.
I finally got it working by switching the Python version I am using.

- - -
When I call `help(User)` with `python==3.8.16`
![image](https://github.com/sqlalchemy/sqlalchemy/assets/41551012/e501f8de-b782-4750-bcf0-6b12fae5b088)

When I use `python==3.9.18`
![image](https://github.com/sqlalchemy/sqlalchemy/assets/41551012/f16b9a8e-6a80-4aac-b178-97a86b803f8c)

Is this intended behavior or is something wrong with my system?
I did use new conda environments and only installed SQLAlchemy so I'd hope this is reproducible.

**Reply by CaselIT** [reply]
> Is this intended behavior or is something wrong with my system?

Don't know, all of this is done by python, it's not something that sqlalchemy can change (at least I don't think it case)

**Reply by CozyDoomer** [reply]
Okay well anyway it is working as intended now with `python>=3.9`, thank you!
Should I create an issue to confirm this doesn't work for `python>=3.7,<3.9` and include it in the documentation?

**Reply by CaselIT** [reply]
> Should I create an issue to confirm this doesn't work for python>=3.7,<3.9 and include it in the documentation?

I'm not sure what you mean. As far as I know this is only ide / linter dependent. At runtime it will run properly in any supported python version

**Reply by CozyDoomer** [reply]
It doesn't seem ide / linter dependent.

I executed this using different python versions:
```python
from sqlalchemy.orm import MappedAsDataclass, DeclarativeBase, mapped_column, Mapped

class Base(MappedAsDataclass, DeclarativeBase):
    """subclasses will be converted to dataclasses"""

class User(Base):
    __tablename__ = "user_account"

    id: Mapped[int] = mapped_column(init=False, primary_key=True)
    name: Mapped[str]

help(User)
```

and it didn't show the attributes for `python>=3.7,<3.9`

**Reply by zzzeek** [reply]
this feature is provided by [pep-681](https://peps.python.org/pep-0681/) so if those older IDEs dont support this pep, it wont work.  that's not really in scope for SQLAlchemy to enumerate what tools do or do not support pep-681, we instead note right at the top of https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#declarative-dataclass-mapping that this feature relies upon pep-681 in order to function, where IDE / type checker support may be limited

**Reply by CozyDoomer** [reply]
I totally get that and it's not something I will push for.
I was only wondering if it's something that should be added to the docs since it's also not working using python's internal `help` function when using older but still supported python versions - independent of IDE.

**Reply by CozyDoomer** [reply]
I will mark the initial dataclasses answer after this last point - thank you for the help!

**Reply by CaselIT** [reply]
I'm not sure what's the reason for this difference.
Plain dataclasses seem to work ok with `help`. I don't think anything different is done in python <3.9 on sqlalchemy side
